midend/interpreter.cpp

```diff
@@ -310,33 +310,49 @@ void SymbolicStruct::dbprint(std::ostream& out) const {
 }
 
 SymbolicHeaderUnion::SymbolicHeaderUnion(const IR::Type_HeaderUnion* type,
-                               bool uninitialized,
-                               const SymbolicValueFactory* factory) :
-        SymbolicStruct(type, uninitialized, factory),
-        valid(new SymbolicBool(false)) {}
+                                         bool uninitialized,
+                                         const SymbolicValueFactory* factory) :
+        SymbolicStruct(type, uninitialized, factory) {}
+
+void SymbolicHeaderUnion::setValid(bool v, cstring field) {
+    if (field) {
+        fieldValue[field]->to<SymbolicHeader>()->setValid(v);
+        for (auto f : type->to<IR::Type_StructLike>()->fields) {
+            if (f->name.name != field)
+                fieldValue[f->name.name]->setAllUnknown();
+        }
+    }
+}
 
-void SymbolicHeaderUnion::setValid(bool v) {
-    if (!v)
-        setAllUnknown();
-    valid = new SymbolicBool(v);
+SymbolicBool* SymbolicHeaderUnion::isValid() const {
+    for (auto f : type->to<IR::Type_StructLike>()->fields) {
+         auto fieldsClone = fieldValue;
+         if (fieldsClone[f->name.name]) {
+             if (fieldsClone[f->name.name]->to<SymbolicHeader>()->valid->value)
+                return new SymbolicBool(true);
+         }
+    }
+    return new SymbolicBool(false);
 }
 
 SymbolicValue* SymbolicHeaderUnion::get(const IR::Node* node, cstring field) const {
-    if (valid->isKnown() && !valid->value)
-        return new SymbolicStaticError(node, "Reading field from invalid header union");
+    auto hu = new SymbolicHeaderUnion(type->to<IR::Type_HeaderUnion>());
```

Comment:
why do you allocate here a new SymbolicHeaderUnion? 
that one will never be valid by construction, so the next statement does nothing.